### PR TITLE
Add Hidden to CursorIcon and CursorSource enums

### LINKED ISSUE #15379: Add CursorIcon::Hidden
## What problem does this solve or what need does it fill?

There are many cases where you may want to hide the mouse cursor. Previously, this could be done by `window.cursor.visible = false`. However, the winit API has recently changed.

Bevy now provides a new abstraction, `CursorIcon`, which allows you to set the cursor shape using entities and components. But this does not provide an option for hiding the cursor - you still need to use the winit API for that.

## What solution would you like?

What I would like is for the CursorIcon enum to have a `Hidden` variant, which indicates that the cursor should not be shown.

## Additional context

In my UI framework, I use the same `CursorIcon` component on entities as well as windows to indicate what cursor shape should be used when hovering over that entity. A system propagates these changes upward: it starts with the current hover entity and walks up the ancestor chain until it finds an entity with a `CursorIcon` component. This component is then copied to the window.

### PR DESCRIPTION
**Pull request by fernanlukban**
# Objective

- Unifies the old winit cursor api (```window.cursor.visible = false```) with the new old winit cursor api (```CursorIcon``` implemented in #14284)
  - Adds a ```Hidden``` enum variant for ```CursorIcon``` and related ```CursorSource``` in the consuming API
- Fixes #15379

## Solution

I just decided to add ```Hidden``` variant to the enum, then decided to fix the errors one by one to find the places where the match guards weren't being exhausted properly. Once I saw that it gets consumed in the ```update_cursors``` function, I decided to see how it gets used to actually change the cursor. Looking at #14284, I saw that there was an intermediary ```CursorSource``` enum that would get used to actually create the cursor. From what I understand, this enum was created becase we can either have system cursors that exist on our OS or a custom one that we define. My thought process then, was to propagate the ```Hidden``` value to this struct and then instead of setting the actual cursor source, figure out how to turn off the visibility. I looked at where the ```window.cursor.visibility``` value that the old winit api was using and then saw that we could set ```winit_window.set_cursor_visible(false)```.

The only thing I'm worried about now is that I'm wondering about if there's a performance hit of doing this way. I haven't really gotten the time to look through the code properly, but previously we were not changing the visibility of this at all. Right now, I am changing the visibility whenever we are updating the cursors. I'm thinking that since this doesn't get changed too often, changing the visibility should just be minimal impact.

The only thing I wanted to do was something along the lines of 

```rust
winit_window.set_cursor_visible(final_cursor.is_some())
```

but I am quite noob at rust and couldn't figure out how to fight back at the borrow checker.

Also, if there are any docs that I should update please let me know.

## Testing

- Built on top of the #14284 example that was updated, added the enum to the variant and ran, working at least for macos, will be able to test on windows when I get back home in two days

https://github.com/user-attachments/assets/550838bb-352b-4d1b-b821-cb316b4b8b8c

---

## Migration Guide

- This might be a breaking change depending on if users ever matched on the ```CursorIcon``` enum, though I'm not really sure if they would ever

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
> I'm thinking that since this doesn't get changed too often, changing the visibility should just be minimal impact.

100% agreed :)

**Review by alice-i-cecile — APPROVED** [review]
Very nice! No complaints here; this is how I would have done it (including chasing errors after adding the variant).

**Review by tychedelia — CHANGES_REQUESTED** [review]
This conflicts with the existing `CursorOptions` field on `Window` that sets the visibility in `changed_windows`. I'm okay with considering the more ECS focused approached, but we should minimally remove `visible` from `CursorOptions`, as well as documenting the platform specific gotchas on the enum here, as well as explore adding a `CursorGrabMode` component and I guess `HitTest` too? Idk, now I'm also like that's too ECS-y lol, but minimally we need to resolve what to do with the existing window settings.

**Review by viridia — APPROVED** [review]

**Comment by eero-lehtinen** [comment]
With this design it becomes harder to toggle visibility. You would have to store the previous cursor somewhere to unhide it later. But it might not matter too much.

I think this option would fit better as a boolean field within the cursor component, but that unfortunately breaks change detection. Could we use the same diffing system that we use for the window settings, where we store a copy of the previous window component and that way get ergonomics of a single unified component and field level change detection?

**Comment by fernanlukban** [comment]
I kind of get the gist of it and agree with the ergonomics. Do you have a pointer of the window settings thing you're referencing?

**Comment by tychedelia** [comment]
> I kind of get the gist of it and agree with the ergonomics. Do you have a pointer of the window settings thing you're referencing?

Look at the `CursorOptions` struct in `window.rs`.

File: crates/bevy_window/src/window.rs [hunk 1]
```diff
@@ -126,7 +126,7 @@ impl NormalizedWindowRef {
 )]
 #[reflect(Component, Default, Debug)]
 pub struct Window {
-    /// The cursor options of this window. Cursor icons are set with the `Cursor` component on the
+    /// The cursor options of this window. Cursor icons are set with the `CursorIcon` component on the
```
Comment (Vrixyz) [inline crates/bevy_window/src/window.rs:129]: 👍  ; I think that's important to note to ease with migration, that being said this property actually refers to `bevy_render`, and it's annoying that we can't link to it through a real link :( (also, should we at all ? 🤔)

Comment (fernanlukban) [reply crates/bevy_window/src/window.rs:129]: no worries! we have had a whole conversation about it inside a discord thread

tl;dr ideally this would live inside of `bevy_winit`, to keep ergonomics the same since most of the use cases regarding icons right now involve `Query<&Window>` and reading the `cursor_options` field which is of the `CursorOptions` type, which originally had `CursorIcon`. Now that we have custom cursor support, this had to live inside `bevy_render` because the custom cursors require access to `Image` which can't be imported inside of `bevy_winit` nor `bevy_window` due to circular dependencies.

though I do agree that the comment should be linked properly - I tried adding the fully qualified name on another comment, but I wasn't sure about how to run the docs to check if it worked locally (i'm still a rust toolchain noob)

File: crates/bevy_winit/src/state.rs [hunk 2]
```diff
@@ -149,6 +149,8 @@ pub struct CustomCursorCache(pub HashMap<CustomCursorCacheKey, winit::window::Cu
 /// A source for a cursor. Is created in `bevy_render` and consumed by the winit event loop.
 #[derive(Debug)]
 pub enum CursorSource {
+    /// A hidden cursor, used to hide the cursor in `winit_window`
+    Hidden,
```
Comment (eero-lehtinen) [inline crates/bevy_winit/src/state.rs:153]: I don't think we should propagate cursor visibility to `bevy_winit`. Custom cursors are required to be changed in such a convoluted way because they require access to the winit event loop. Cursor visibility can be changed directly with `NonSendMut<WinitWindows>` in any system just like it was done before with `cursor_options`.

**Review by eero-lehtinen — COMMENTED** [review]
I'm certainly in favor of moving cursor related options to the cursor component, but it feels weird losing the ability that winit provides to separately toggle visibility and choose the displayed icon. I don't have super strong feelings about though, this way is ok too. 

The change to automatically add the cursor component to every window is good. I should have done it that way before. Though I think observers would work better in avoiding scheduling issues.

File: crates/bevy_window/src/window.rs (hunk 1 above)
Comment (Vrixyz) [reply crates/bevy_window/src/window.rs:129]: thanks for the insight ! 

> I tried adding the fully qualified name on another comment

Nice try! the CI docs step fails due to that. Check it out (and I recommend looking at the .github workflow file to see more details on what commands are run :+1:

I fear it's impossible to link because bevy_window doesn't rely on bevy_render :(

well I imagine we theoretically could by adding the dependency if building docs, but I personally wouldn't approve that :/

<details><summary>Continued rambling about bringing CursorIcon back to `bevy_window`</summary>
<p>

I'd prefer a `Custom` empty enum variant, and let the other crates handle their custom thing however they want 🤔.

Or a generic Custom<T> ? or a Box with a dyn trait something inside ? 

</p>
</details> 

> I wasn't sure about how to run the docs

`cargo doc --open` will bring you far enough 🤞

File: crates/bevy_winit/src/state.rs (hunk 2 above)
Comment (fernanlukban) [reply crates/bevy_winit/src/state.rs:153]: I'm starting to think what we really need to do is do something along the lines of not having a cursor from being instead `Hidden` to just not having a cursor source or `None`. This clearly denotes the feature set each cursor source has - the `SystemCursor` is heavily part of the `winit` package we use, hence the related feature set. Now that we have `CursorSource::Custom`, if we think that over time it ends up having a different feature set, maybe something like if we were to support gifs as cursors, or sprite sheets as cursors with animations, there might be options relating how fast to play the gif or whether or not to loop it's animation. This way, most of the examples will look almost the same in functionality, just changing a bool

```
pub enum CursorSource {
    None,
    Custom{ custom_cursor: CustomCursor, visibility: bool },
    System{ system_cursor: SystemCursor, visibility: bool },
}
```

Comment (fernanlukban) [reply crates/bevy_winit/src/state.rs:153]: As to the 

> Cursor visibility can be changed directly with `NonSendMut<WinitWindows>` in any system just like it was done before with `cursor_options`.

I don't get quite this means? Are you suggesting that if it isn't a `CustomCursor` then we should just use the system `winit` related API directly? I think we were initially against that because we would have two API's and it could get confusing, so I would like your take on it.

File: crates/bevy_render/src/view/window/cursor.rs [hunk 3]
```diff
@@ -29,7 +28,7 @@ impl Plugin for CursorPlugin {
         app.register_type::<CursorIcon>()
             .init_resource::<CustomCursorCache>()
             .add_systems(Last, update_cursors)
-            .add_systems(Update, add_cursor_to_windows);
+            .observe(add_cursor_to_windows);
```
Comment (fernanlukban) [inline crates/bevy_render/src/view/window/cursor.rs:31]: @eero-lehtinen this was quite pretty to do, I remember reading the example for this last week and now this seems so nice

File: examples/input/mouse_grab.rs [hunk 4]
```diff
@@ -19,9 +19,7 @@ fn grab_mouse(
     // There are cases where we can have multiple windows and cursors,
     // but not in this example. Keeping the naming convention since it
     // exists in other examples
-    let Ok((mut window, mut cursor)) = windows_and_cursors.get_single_mut() else {
```
Comment (fernanlukban) [inline examples/input/mouse_grab.rs]: it ended up fixing the non-determinism thing here so that we will always have the cursor when windows are created which is perfect

**Review by viridia — APPROVED** [review]

File: crates/bevy_winit/src/state.rs [hunk 5]
```diff
@@ -154,6 +154,8 @@ pub struct CustomCursorCache(pub HashMap<CustomCursorCacheKey, winit::window::Cu
...
```
Comment (tychedelia) [inline crates/bevy_winit/src/state.rs:158]: This is still missing documentation about platform specific behavior.

File: crates/bevy_winit/src/state.rs (hunk 2 above)
Comment (eero-lehtinen) [reply crates/bevy_winit/src/state.rs:158]: What I would be proposing is to just move cursor options from the window component to the cursor component and doing absolutely everything the exact same it was done before and make a new cursor component.
```rust
struct Cursor {
    pub visible: bool,
    ... other options ...
    pub icon: CursorIcon,
}
```
Then move everything that was done in the window update system to the cursor update system. 

But I guess that is a different proposal than what is included in the issue, so we can also implement this proposal if others like it better.


> I don't get quite this means? Are you suggesting that if it isn't a `CustomCursor` then we should just use the system `winit` related API directly? I think we were initially against that because we would have two API's and it could get confusing, so I would like your take on it.

Yes. `CursorSource` is an implementation detail for smuggling cursor icon bytes to the `bevy_winit` crate. It would lead to less spread out code and less complexity to just set the `WinitWindows` cursor visibility boolean in the `bevy_render` cursor system. It doesn't create any additional APIs. Functionally your code does the same thing so consider what would be simpler.

Comment (eero-lehtinen) [reply crates/bevy_winit/src/state.rs:158]: > Now that we have CursorSource::Custom, if we think that over time it ends up having a different feature set.

`CursorSource` is a direct equivalent to `winit`s native cursor sources. Those are clearly defined and not flexible, so the `CursorSource` doesn't need to be flexible.

E.g. if we want to use sprite sheets as animated cursors, those would need to be converted to be `winit` compatible outside of `bevy_winit` anyway because in can't use assets.
